lib/src/commands/generate.dart

```diff
@@ -0,0 +1,188 @@
+    final path = argResults?['path'];
+    final url = argResults?['url'];
+    final String destination = argResults!['destination'];
+    final String packageName = argResults!['package_name'];
+    final bool force = argResults!['force'];
+
+    final bool generateServer = argResults!['server'];
+    final bool generateClient = argResults!['client'];
+    if (!generateServer && !generateClient) {
+      _logger.err('You must specify at least one of --server or --client');
+      return ExitCode.usage.code;
+    }
+
+    final destDirectory = Directory(destination);
+    if (destDirectory.existsSync() && !force) {
+      _logger.err(
+          'The destination directory already exists. Use --force to overwrite');
+      return ExitCode.usage.code;
+    } else if (destDirectory.existsSync()) {
+      destDirectory.deleteSync(recursive: true);
+    }
+
+    final String rawSchemaText;
+    final String fileExtension;
+
+    if (path != null && url != null) {
+      _logger.err('You can only specify a path or a URL, not both');
+      return ExitCode.usage.code;
+
+      // Read the OpenAPI schema from the specified path
+    } else if (path != null) {
+      final f = File(path);
+      if (f.existsSync()) {
+        _logger.info('Path to the OpenAPI schema file: $path');
+        rawSchemaText = f.readAsStringSync();
+        fileExtension = p.extension(path);
+      } else {
+        _logger.err('The file at the specified path does not exist');
+        return ExitCode.usage.code;
+      }
+
+      // Read the OpenAPI schema from the specified URL
+    } else if (url != null) {
+      final Uri uri;
+      final http.Response response;
+      try {
+        uri = Uri.parse(url);
+      } on FormatException catch (e) {
+        _logger.err('The URL is not valid: $e');
+        return ExitCode.usage.code;
+      }
+      _logger.info('URL to the OpenAPI schema file: $url');
+
+      try {
+        response = await http.get(uri);
+        if (response.statusCode != 200) {
+          _logger.err('The URL is not reachable: ${response.reasonPhrase}');
+          return ExitCode.usage.code;
+        }
+      } on Exception catch (e) {
+        _logger.err('The URL is not reachable: $e');
+        return ExitCode.usage.code;
+      }
+      rawSchemaText = response.body;
+      fileExtension = p.extension(url);
+    } else {
+      _logger.err('You must specify a path or a URL');
+      return ExitCode.usage.code;
+    }
+
+    final schema =
+        parseRawSchemaFromString(rawSchemaText, fileExtension, _logger);
+
+    if (schema == null) {
+      _logger.err('The content of the file is not a valid JSON or YAML');
+      return ExitCode.usage.code;
+    }
+
+    if (schema.containsKey("swagger")) {
+      _logger.err('The OpenAPI version 2.0 is not supported');
+      return ExitCode.usage.code;
+    }
+
+    var spec = OpenApi.fromJson(schema);
+    final centralizer = SchemaCentralizer(spec);
+    spec = centralizer.centralizedSpec();
```

Comment:
> I respectfully disagree with the approach of not generating data models from the definitions found in the paths

If the API author wanted to create a reusable data model, that would have done so in their API specification and it would show up in the components.

> The reality is that many developers rely on backend frameworks or third-party APIs to generate the OpenAPI spec, rather than writing it from scratch.

That is correct. FastAPI in Python is a great option and there are plenty of examples where request/response data models can/should be specified by the author. Again, the API author can define this. It seems that if it is missing from the spec (as a pre-defined component), it is intentional.

> No other major generator takes the approach you described,

What is the output of the Dart [openapi-generator](https://github.com/OpenAPITools/openapi-generator) for the same spec you provided? Did it provide a typed response?

> I assumed that this would be opt-out.

Due to previous described reasons in other comments, this is difficult to get right because the component is not explicitly defined.

> DevX and popularity will suffer greatly.

"suffer greatly"? That might be a stretch. All I am saying is to add a `centralize=true` flag option. Again, look at the [openapi-generator](https://github.com/OpenAPITools/openapi-generator) options and tell me how that dev experience is. Automatic code generation comes with nuances that there need to be flags in place. There is no once size fits all.

Reply:
Just tried it, it creates a model for the request.

Here is a schema with no components, everything is declared in `paths`.
<details><summary>Schema</summary>
<p>

```
openapi: 3.1.0
info:
  title: Swagger Petstore - OpenAPI 3.1
  contact:
    email: apiteam@swagger.io
  license:
    name: Apache 2.0
    url: http://www.apache.org/licenses/LICENSE-2.0.html
  version: 1.0.11
externalDocs:
  description: Find out more about Swagger
  url: http://swagger.io
servers:
  - url: https://petstore3.swagger.io/api/v3
paths:
  /pet:
    put:
      tags:
        - pet
      summary: Update an existing pet
      description: Update an existing pet by Id
      operationId: updatePet
      parameters:
        - name: status
          in: query
          description: Status values that need to be considered for filter
          schema:
            required:
              - name
              - photoUrls
            type: object
            properties:
              id:
                type: integer
                format: int64
                examples: [10]
              name:
                type: string
                examples: [doggie]
              photoUrls:
                type: array
                xml:
                  wrapped: true
                items:
                  type: string
                  xml:
                    name: photoUrl
              status:
                type: string
                description: pet status in the store
                enum:
                  - available
                  - pending
                  - sold
      requestBody:
        description: Update an existent pet in the store
        content:
          application/json:
            schema:
              required:
                - name
                - photoUrls
              type: object
              properties:
                id:
                  type: integer
                  format: int64
                  examples: [10]
                name:
                  type: string
                  examples: [doggie]
                photoUrls:
                  type: array
                  xml:
                    wrapped: true
                  items:
                    type: string
                    xml:
                      name: photoUrl
                status:
                  type: string
                  description: pet status in the store
                  enum:
                    - available
                    - pending
                    - sold

        required: true
      responses:
        '200':
          description: Successful operation
          content:
            application/json:
              schema:
                required:
                  - name
                  - photoUrls
                type: object
                properties:
                  id:
                    type: integer
                    format: int64
                    examples: [10]
                  name:
                    type: string
                    examples: [doggie]
                  photoUrls:
                    type: array
                    xml:
                      wrapped: true
                    items:
                      type: string
                      xml:
                        name: photoUrl
                  status:
                    type: string
                    description: pet status in the store
                    enum:
                      - available
                      - pending
                      - sold
``` 

</p>
</details> 


It generates a model for it `update_pet_request.dart`
```dart
//
// AUTO-GENERATED FILE, DO NOT MODIFY!
//

// ignore_for_file: unused_element
import 'package:built_collection/built_collection.dart';
import 'package:built_value/built_value.dart';
import 'package:built_value/serializer.dart';

part 'update_pet_request.g.dart';

/// UpdatePetRequest
///
/// Properties:
/// * [id] 
/// * [name] 
/// * [photoUrls] 
/// * [status] - pet status in the store
@BuiltValue()
abstract class UpdatePetRequest implements Built<UpdatePetRequest, UpdatePetRequestBuilder> {
  @BuiltValueField(wireName: r'id')
  int? get id;

  @BuiltValueField(wireName: r'name')
  String get name;

  @BuiltValueField(wireName: r'photoUrls')
  BuiltList<String> get photoUrls;

  /// pet status in the store
  @BuiltValueField(wireName: r'status')
  UpdatePetRequestStatusEnum? get status;
  // enum statusEnum {  available,  pending,  sold,  };

  UpdatePetRequest._();

  factory UpdatePetRequest([void updates(UpdatePetRequestBuilder b)]) = _$UpdatePetRequest;

  @BuiltValueHook(initializeBuilder: true)
  static void _defaults(UpdatePetRequestBuilder b) => b;

  @BuiltValueSerializer(custom: true)
  static Serializer<UpdatePetRequest> get serializer => _$UpdatePetRequestSerializer();
}

class _$UpdatePetRequestSerializer implements PrimitiveSerializer<UpdatePetRequest> {
  @override
  final Iterable<Type> types = const [UpdatePetRequest, _$UpdatePetRequest];

  @override
  final String wireName = r'UpdatePetRequest';

  Iterable<Object?> _serializeProperties(
    Serializers serializers,
    UpdatePetRequest object, {
    FullType specifiedType = FullType.unspecified,
  }) sync* {
    if (object.id != null) {
      yield r'id';
      yield serializers.serialize(
        object.id,
        specifiedType: const FullType(int),
      );
    }
    yield r'name';
    yield serializers.serialize(
      object.name,
      specifiedType: const FullType(String),
    );
    yield r'photoUrls';
    yield serializers.serialize(
      object.photoUrls,
      specifiedType: const FullType(BuiltList, [FullType(String)]),
    );
    if (object.status != null) {
      yield r'status';
      yield serializers.serialize(
        object.status,
        specifiedType: const FullType(UpdatePetRequestStatusEnum),
      );
    }
  }

  @override
  Object serialize(
    Serializers serializers,
    UpdatePetRequest object, {
    FullType specifiedType = FullType.unspecified,
  }) {
    return _serializeProperties(serializers, object, specifiedType: specifiedType).toList();
  }

  void _deserializeProperties(
    Serializers serializers,
    Object serialized, {
    FullType specifiedType = FullType.unspecified,
    required List<Object?> serializedList,
    required UpdatePetRequestBuilder result,
    required List<Object?> unhandled,
  }) {
    for (var i = 0; i < serializedList.length; i += 2) {
      final key = serializedList[i] as String;
      final value = serializedList[i + 1];
      switch (key) {
        case r'id':
          final valueDes = serializers.deserialize(
            value,
            specifiedType: const FullType(int),
          ) as int;
          result.id = valueDes;
          break;
        case r'name':
          final valueDes = serializers.deserialize(
            value,
            specifiedType: const FullType(String),
          ) as String;
          result.name = valueDes;
          break;
        case r'photoUrls':
          final valueDes = serializers.deserialize(
            value,
            specifiedType: const FullType(BuiltList, [FullType(String)]),
          ) as BuiltList<String>;
          result.photoUrls.replace(valueDes);
          break;
        case r'status':
          final valueDes = serializers.deserialize(
            value,
            specifiedType: const FullType(UpdatePetRequestStatusEnum),
          ) as UpdatePetRequestStatusEnum;
          result.status = valueDes;
          break;
        default:
          unhandled.add(key);
          unhandled.add(value);
          break;
      }
    }
  }

  @override
  UpdatePetRequest deserialize(
    Serializers serializers,
    Object serialized, {
    FullType specifiedType = FullType.unspecified,
  }) {
    final result = UpdatePetRequestBuilder();
    final serializedList = (serialized as Iterable<Object?>).toList();
    final unhandled = <Object?>[];
    _deserializeProperties(
      serializers,
      serialized,
      specifiedType: specifiedType,
      serializedList: serializedList,
      unhandled: unhandled,
      result: result,
    );
    return result.build();
  }
}

class UpdatePetRequestStatusEnum extends EnumClass {

  /// pet status in the store
  @BuiltValueEnumConst(wireName: r'available')
  static const UpdatePetRequestStatusEnum available = _$updatePetRequestStatusEnum_available;
  /// pet status in the store
  @BuiltValueEnumConst(wireName: r'pending')
  static const UpdatePetRequestStatusEnum pending = _$updatePetRequestStatusEnum_pending;
  /// pet status in the store
  @BuiltValueEnumConst(wireName: r'sold')
  static const UpdatePetRequestStatusEnum sold = _$updatePetRequestStatusEnum_sold;

  static Serializer<UpdatePetRequestStatusEnum> get serializer => _$updatePetRequestStatusEnumSerializer;

  const UpdatePetRequestStatusEnum._(String name): super(name);

  static BuiltSet<UpdatePetRequestStatusEnum> get values => _$updatePetRequestStatusEnumValues;
  static UpdatePetRequestStatusEnum valueOf(String name) => _$updatePetRequestStatusEnumValueOf(name);
}


```

Reply:
> All I am saying is to add a centralize=true flag option.  

This doesn't mean anything to someone who doesnt understand the OpenAPI spec, which is most people.  
I name the flag `just-work-correctly`, in which case it would be kinda funny if it wasn't on by default. 😂

Reply:
Man... that is ugly generated code.

> Just tried it, it creates a model for the request.

Ok cool. I am not completely opposed to it being opt-in or opt-out, it just should be optional. The default can go either way.

Can you rebase your branch on the latest `main`?

Reply:
I won't die on this hill, but as long as it documented very clearly in the docs it's ok I guess.

**To Generate Models for All Objects, use `--centralize`**

Reply:
> This doesn't mean anything to someone who doesnt understand the OpenAPI spec, which is most people.
I name the flag just-work-correctly, in which case it would be kinda funny if it wasn't on by default.

Facts. You are starting to sell me @dickermoshe

Reply:
You did 99% of the work on this package, It's your call

Reply:
Rebased.